dsmr_parser/objects.py

```diff
@@ -24,24 +20,55 @@ class Telegram(object):
-        self._telegram_specification = telegram_specification
-        self._telegram_parser = telegram_parser
-        self._obis_name_mapping = dsmr_parser.obis_name_mapping.EN
-        self._reverse_obis_name_mapping = dsmr_parser.obis_name_mapping.REVERSE_EN
-        self._dictionary = self._telegram_parser.parse(telegram_data)
-        self._item_names = self._get_item_names()
-
-    def __getattr__(self, name):
-        ''' will only get called for undefined attributes '''
-        obis_reference = self._reverse_obis_name_mapping[name]
-        value = self._dictionary[obis_reference]
-        setattr(self, name, value)
-        return value
-
-    def _get_item_names(self):
-        return [self._obis_name_mapping[k] for k, v in self._dictionary.items()]
+    def __init__(self, *args, **kwargs):
+        self._mbus_channel_devices = {}
+        self._item_names = []
+        super().__init__(*args, **kwargs)
+
+    def add(self, obis_reference, dsmr_object):
+        # Update name mapping used to get value by attribute. Example: telegram.P1_MESSAGE_HEADER
+        # Also keep track of the added names internally
+        obis_name = obis_name_mapping.EN[obis_reference]
+        setattr(self, obis_name, dsmr_object)
+        if obis_name not in self._item_names:  # TODO solve issue with repeating obis references
+            self._item_names.append(obis_name)
+
+        # Group Mbus related values into a MbusDevice object.
+        # TODO MaxDemandParser (BELGIUM_MAXIMUM_DEMAND_13_MONTHS) returns a list
+        if isinstance(dsmr_object, DSMRObject) and dsmr_object.is_mbus_reading:
+            self._add_mbus(obis_reference, dsmr_object)
+
+        # Fill dict which is only used for backwards compatibility
+        if obis_reference not in self:
+            self[obis_reference] = dsmr_object
+
+    def _add_mbus(self, obis_reference, dsmr_object):
+        """
+        The given DsmrObject is assumed to be Mbus related and will be grouped into a MbusDevice.
+        Grouping is done by the DsmrObject channel ID.
+        """
+        channel_id = dsmr_object.obis_id_code[1]
+
+        # Create new MbusDevice for the first record on this channel_id
+        if channel_id not in self._mbus_channel_devices:
+            self._mbus_channel_devices[channel_id] = MbusDevice(channel_id=channel_id)
+
+        mbus_device = self._mbus_channel_devices[channel_id]
+        mbus_device.add(obis_reference, dsmr_object)
+
+    @property
+    def MBUS_DEVICES(self):
+        """ Return MbusDevice objects which are used for water, heat and gas meters. """
+        return list(self._mbus_channel_devices.values())
+
+    def get_mbus_device_by_channel(self, channel_id):
+        """
+        :rtype: MbusDevice|None
+        """
+        return self._mbus_channel_devices.get(channel_id)
+
+    def __len__(self):
+        return len(self._item_names)
 
     def __iter__(self):
         for attr in self._item_names:
```

Comment:
TODO: yield MBUS_DEVICES here

Reply:
What about: 
- adding MBUS_DEVICES to _item_names?
- adding MBUS_DEVICES to the object with `setattr`, if it not already added?

I am strugling with the meaning of __len__ here. How do you meaningfully express the length of a telegram object, in a way that it is also useful and predictable for users of the object?

Reply:
Thanks, i've processed your feedback. Len is removed again, because it was related to trying to maintain backwards compatibility with the dict. This is simplified with the inheritance of dict. 

Telegram should replace the goal of the dict, but deprecating and moving to 2.0.0 is something for a separate ticket